### PR TITLE
*: Change privilege to use Auth* identity

### PR DESCRIPTION
**Pull request by morgo**
### What problem does this PR solve? 

Commands like `SHOW GRANTS` or `SHOW GRANTS FOR CURRENT_USER()` now work as in MySQL.  Previously they failed to match a user, because they were looking for an exact match of username/hostname in the privileges table, and not applying the wildcard/matching rules that occur.

This partially addresses https://github.com/pingcap/tidb/issues/7614

### What is changed and how it works?

When available, use Auth* portion of identity (i.e. the cached entries in the privileges table).
Fix parser use of current_user() to add the Auth* credentials.

### Check List 

Tests 

 - Unit test
 - Manual test (add detailed scripts or steps below)

Code changes

 - Has interface methods change

The signature doesn't change, but it might seem a bit 'magic' how the Auth* identity params are now prefered over the regular ones.

Side effects

 - Possible performance regression
 - Increased code complexity

There is now additional logic in checking if a user matches privileges.

Related changes

I don't think it requires docs.  It is just a bug fix.

### CODE REVIEW

File: planner/core/planbuilder.go [hunk 1]
```diff
@@ -1615,6 +1615,12 @@ func buildShowSchema(s *ast.ShowStmt) (schema *expression.Schema) {
 		names = []string{"Database", "Create Database"}
 	case ast.ShowGrants:
 		names = []string{fmt.Sprintf("Grants for %s", s.User)}
```
Comment (zz-jason) [inline planner/core/planbuilder.go]: this line should be removed.

**Comment by zz-jason** [comment]
@tiancaiamao PTAL

File: planner/core/planbuilder.go [hunk 2]
```diff
@@ -1615,7 +1615,12 @@ func buildShowSchema(s *ast.ShowStmt) (schema *expression.Schema) {
 	case ast.ShowCreateDatabase:
 		names = []string{"Database", "Create Database"}
 	case ast.ShowGrants:
-		names = []string{fmt.Sprintf("Grants for %s", s.User)}
+		if s.User != nil {
+			names = []string{fmt.Sprintf("Grants for %s", s.User)}
+		} else {
+			// Don't know the name yet, so just say "user"
+			names = []string{fmt.Sprintf("Grants for User")}
```
Comment (zz-jason) [inline planner/core/planbuilder.go]: we can save the function call of `fmt.Sprintf()`, use the string "Grants for User" instead.

File: privilege/privileges/privileges.go [hunk 3]
```diff
@@ -141,5 +141,9 @@ func (p *UserPrivileges) UserPrivilegesTable() [][]types.Datum {
 // ShowGrants implements privilege.Manager ShowGrants interface.
 func (p *UserPrivileges) ShowGrants(ctx sessionctx.Context, user *auth.UserIdentity) ([]string, error) {
 	mysqlPrivilege := p.Handle.Get()
+	if len(user.AuthUsername) > 0 && len(user.AuthHostname) > 0 {
+		return mysqlPrivilege.showGrants(user.AuthUsername, user.AuthHostname), nil
+	}
 	return mysqlPrivilege.showGrants(user.Username, user.Hostname), nil
+
```
Comment (zhexuany) [inline privilege/privileges/privileges.go]: remove this extra space.

**Pull request merged**